### ISSUE TITLE
Add additional param to inputfield::render() to overwrite markup

### ISSUE THREAD

**Issue by pwFoo** [processwire/processwire-requests#66]
Would be great to overwrite default inputfield markup for example with an optional parameter to render() method.

Markup is hardcoded (example radio input):
```
$out .= 
				"<li$liAttr><label$attrs>" . 
				"<input$checked$disabled " . 
				"type='radio' " . 
				"name='{$this->name}' " . 
				"id='$id' " .
				"value='" . htmlspecialchars($key, ENT_QUOTES) . "' />" . 
				"<span class='$textClass'>$label</span>" . 
				"</label></li>";
		}
```

Markup overwrite would help to support css frameworks which needs markup order like that:
```
<input type='checkbox' " . $field->getAttributesString($attrs) . " />
<label for='{$field->id}'>$label</label>
```

**Comment by pwFoo** [comment]
@ryancramerdesign 
The markup order input before label is needed to style label dependent on input checkbox / radio is checked or unchecked. It isn't possible to style the parent of an inputfield by css selector.

**Comment by teppokoivula** [comment]
There are modules that alter the look and/or behaviour of existing inputfields with JavaScript or CSS and those modules may depend on a specific markup. If we make changing it entirely (compared to, for an example, enabling one to add her own classes, attributes, etc.) this may cause issues with such modules.

In theory it's as simple as explaining to users that "modules A and B don't work together, deal with it", but I think it's still a valid concern and definitely something to keep in mind when overriding built-in inputfield markup.

**Comment by pwFoo** [comment]
@teppokoivula 
Inputfield frontend usage the backend depencencies shouldn't be a problem I think. And it would make it easier to just use PW fields / form api to handle forms. 

Would You prefer to hook the existing inputfields or create new inputfields for frontend usage?

**Comment by teppokoivula** [comment]
Right, thanks for pointing out that this was intended mostly for front-end use.

When inputfields were mentioned, I (wrongly) assumed it was intended for backend use. Front-end forms using built-in inputfields are something I use very rarely myself, partially for the reason you've mentioned here, i.e. because they're not configurable enough, and partially because I tend to turn to FormBuilder instead for most of my form-related needs :)

Taking this point into account somewhat waters down the concerns I expressed above, and your original approach (optional param provided for the render method) makes a lot of sense to me.

**Comment by ivangretsky** [comment]
This would really be awesome! A big step to make one of the most attractive features for the newcomers from the world of Wordpress and Joomla - the so called CCK - complete. The ability to build client forms with the same instruments. For now we sometimes have to use other tools (like suggested [here](https://processwire.com/talk/topic/13798-forms-engine/?do=findComment&comment=124126) by someone respected in the community, for example) instead of powerful in-the-core methods just because we can't style the inputfields the way client or visual designer wants to see them.

This question is [one of the most discussed](https://processwire.com/talk/topic/2089-create-simple-forms-using-api/?page=1) - so it must be important.

Maybe some kind of mini-templates could be used here?

**Comment by jmartsch** [comment]
I would really love to see this, as some Frameworks (mainly Material Design based ones) need the label after the input. I am also talking about frontend use here.
I even got a case with a material design where the input has to be inside the label:
```
<div class="group">
                <label>
                    <input type="text" name="Nachname" required value="{$sanitizer->text($input->post->Nachname)}">
                    <span class="form-error">{$strings.form_error_required}</span>
                    <span class="form-label">{$strings.surname}</span>
                </label>
            </div>
```
So it would be nice if we could customize the order or placement of the items. Maybe something like this:
```
$form_markup = array(
        'list' => "<div {attrs}>{out}</div>",
        'item_template' => "<div {attrs}>{theinput}{description}{error}{notes}{item_label}</label></div>",
        'item_label' => "<span class='form-label'>{out}</span>"
);
```

**Comment by pwFoo** [comment]
At the moment form api is still designed for backend usage.
Is there a form framework which could be easily integrated in PW to replace the form api inputfields?
